### PR TITLE
Add more GSI test variables, test references, and observation bias group to gsi-ncdiag

### PR DESCRIPTION
**Pull request by emilyhcliu**
This PR has the following update:

- Add observation bias groups into obs file for radiance type
- Add more test data and test reference for the radiances data type
- Add handling to GSI related test variables, test references, and observation bias groups.  Addition of these variables are optional
- Recover the feature of using "dynamic" sample size for subset files

For conventional data, the "nrecs" was used to determine the number of samples in the subset files, so the sample size is larger for larger data sets (representativeness).  However, this feature was gone since the "nrecs" was removed.  So, fixed sample size is used (100).  This is not good.  I think the "dynamic" sample size was a very nice feature (thanks to Cory).   So, I added the feature of using "dynamic" sample size for conventional data back.  The "nrecs" is obtained by counting the unique members in the record_number vector.

- Update AMSUA output file 

I am having trouble upload (git push) the update AMSUA file to Git repository since yesterday. 
     
Uploading LFS objects:   0% (0/1), 3.4 KB | 0 B/s  ....     The process was hanging until time out  :-( 

The error message is:

Post https://lfs.github.com/JCSDA/ioda-converters/objects/269801067216b3da5c8a064dce88e8bc5c5e1cb3fb75235d63e884d47a8a573c/verify: dial tcp 140.82.112.5:443: i/o timeout
error: failed to push some refs to 'https://github.com/JCSDA/ioda-converters'

It is not a network problem.  I can clone, push, pull, commit, ...etc with Git repository without problems  Any suggestion?  Is there an alternative way to get this file up to Github?

### CODE REVIEW

**Comment by emilyhcliu** [comment]
The test failed because the updated AMSU-A output file has not been pushed to the branch yet.

**Comment by srherbener** [comment]
@emilyhcliu, do you have the following lines in your `$HOME/.gitconfig `file:
```
[filter "lfs"]
        clean = git-lfs clean -- %f
        smudge = git-lfs smudge -- %f
        process = git-lfs filter-process
        required = true
```

If not, that might be something to try to fix the git-lfs problem. By the way, running ``git lfs install`` should place these lines in the `$HOME/.gitconfig` file. But, you only need to run ``git lfs install`` once after installing git-lfs on your system.

**Comment by emilyhcliu** [comment]
@srherbener I do have the lines related lfs you mentioned in my .gitconfig file.  I just tried to push the AMSUA file again, and the uploading process is till hanging.

**Comment by srherbener** [comment]
@emilyhcliu, sorry that didn't work.

@mmiesch, could you help us get this git-lfs issue resolved? Thanks!

**Comment by mmiesch** [comment]
git-lfs can be configured to adjust how long it waits to time out  - see here - maybe one or more of these settings can help?

https://github.com/git-lfs/git-lfs/blob/master/docs/man/git-lfs-config.5.ronn#L32

**Comment by emilyhcliu** [comment]
@mmiesch Thanks for the information  I will try these options.

**Comment by emilyhcliu** [comment]
@srherbener @mmiesch I push the obs file from S4 successfully.

File: src/gsi-ncdiag/proc_gsi_ncdiag.py.in [hunk 1]
```diff
@@ -99,6 +99,12 @@ ap.add_argument("-g", "--geovals_dir",
                 help="Path to directory to output observations")
 ap.add_argument("-d", "--obsdiag_dir",
                 help="Path to directory to output observations")
+ap.add_argument("-obsbias", "--add_obsbias", default=False,
```
Comment (CoryMartin-NOAA) [inline src/gsi-ncdiag/proc_gsi_ncdiag.py.in]: I think Unix convention is that "--add_obsbias" is fine but the "-" argument should just be one letter either lowercase or capital. Can we shorten the first arguments to be shorter?

Comment (CoryMartin-NOAA) [reply src/gsi-ncdiag/proc_gsi_ncdiag.py.in]: I'd suggest, 'b' for obsbias, 'q' for qcvars, and t for testvars. In reality, users will probably use the '--' arguments but these shorter ones will be helpful for quicker typing on the command line.

File: src/gsi-ncdiag/subset_files.py [hunk 2]
```diff
@@ -16,30 +16,46 @@ def subset(infile, nlocsout, suffix, geofile, diagfile):
...
+    # Get number of records (nrecsin)
+    recnum = ncin.variables['record_number@MetaData'][:]
+    recnum_uniq = np.unique(recnum)
+    nrecsin = len(recnum_uniq)
     nlocsin = len(ncin.dimensions['nlocs'])
     nvars = len(ncin.dimensions['nvars'])
-    nsamples = nlocsout
-    npossible = nlocsin
+    if nrecsin > 100.:
+        nsamples = int(-(-nlocsout/10.))  # just picked 10 randomly
```
Comment (CoryMartin-NOAA) [inline src/gsi-ncdiag/subset_files.py]: why negative negative?

File: src/gsi-ncdiag/test_gsidiag.py.in [hunk 3]
```diff
@@ -20,10 +20,22 @@ parser.add_argument('-t', '--type',
 parser.add_argument('-p', '--platform',
                     help="platform to run, only works on conv",
                     type=str)
+parser.add_argument('-obsbias', '--add_obsbias',
```
Comment (CoryMartin-NOAA) [inline src/gsi-ncdiag/test_gsidiag.py.in]: Same comment/suggestions as before. The - args shouldn't be that long if possible.

File: src/gsi-ncdiag/subset_files.py (hunk 2 above)
Comment (emilyhcliu) [reply src/gsi-ncdiag/subset_files.py]: @CoryMartin-NOAA This is your original coding.  You are right,  the double negative can be removed.

Comment (CoryMartin-NOAA) [reply src/gsi-ncdiag/subset_files.py]: Ha! Well I don't know what I was thinking here originally then.

Comment (emilyhcliu) [reply src/gsi-ncdiag/subset_files.py]: @CoryMartin-NOAA I probably would do the same thing (thinking about the math, not the coding)  your original design about setting the sample size is a very good thing to keep!!

File: src/gsi-ncdiag/test_gsidiag.py.in (hunk 3 above)
Comment (emilyhcliu) [reply src/gsi-ncdiag/test_gsidiag.py.in]: @CoryMartin-NOAA Will do!!

File: src/gsi-ncdiag/test_gsidiag.py.in [hunk 4]
```diff
@@ -20,10 +20,22 @@ parser.add_argument('-t', '--type',
...
                     type=str)
+parser.add_argument('-b', '--add_obsbias',
+                    help="add obsbias groups, only works on rad",
+                    type=str)
+parser.add_argument('-q', '--add_qcvars',
+                    help="add qc variables, only works on rad",
+                    type=str)
+parser.add_argument('-r', '--add_testrefs',
+                    help="add test variables, only works on rad",
+                    type=str)
```
Comment (emilyhcliu) [inline src/gsi-ncdiag/test_gsidiag.py.in]: Change  '**-t'**,     '--add_testvars'    to '**-r**',    '--add_testrefs' due to conflict of using '-t '
( '-t' is already used by data type )

**Comment by emilyhcliu** [comment]
@CoryMartin-NOAA Just summary what I modified according to your suggestion:

(1) Shortened arguments 

    -obsbias to -o 
    -qcvars to -q
    -testrefs to -r

I did not use -t for testrefs because -t is already used for file type
The test_gsidiag.py.in checks if the arguments are unique

(2) Removed the double negatives:

       int ( -(-nrecs/10) ) ---> int( nrecs/10 ) 

Please let me know if you have more thoughts.  Thanks for reviewing this PR.

**Pull request merged**